pkg/sentry/platform/kvm/bluepill_amd64.s

```diff
@@ -39,18 +29,8 @@
 
 // See bluepill.go.
 TEXT ·bluepill(SB),NOSPLIT,$0
-begin:
 	MOVQ vcpu+0(FP), AX
-	LEAQ VCPU_CPU(AX), BX
 	BYTE CLI;
-check_vcpu:
-	MOVQ CPU_SELF(GS), CX
-	CMPQ BX, CX
-	JE right_vCPU
-wrong_vcpu:
-	CALL ·redpill(SB)
-	JMP begin
-right_vCPU:
 	RET
 
```

Comment:
the reason to remove the check:

    m.Get() has guaranteed that if any OS thread TID is in guest,
    m.vCPUs[TID] points to the vCPU in which the OS thread TID is running.

    So if m.Get() returns with the corrent context in guest,
    the vCPU of it must be the same as what Get() returns.

    So bluepill() doesn't need to check if the vCPU is matched or not.
    The check need to access to %gs register which will not points
    to vCPU later when KPTI for gvisor is enabled. We can still
    fetch the vCPU pointer from %gs later (when %gs points to kernelEntry),
    but it needs the ENTRY_CPU_SELF which is generated by
    ring0/offset_amd64.go. So we just simply remove the check.

Reply:
Okay, I believe it but I will need to think about it a bit :)

Reply:
Is there any easy way to use ENTRY_CPU_SELF in kvm package? Or we can implement the bluepill() in ring0. Or we can add a GetCPUFromGS() in ring0, and call it and check it after bluepill() returns?

Reply:
And I'm planing to use percpu trampoline in the entry code, and it can get the percpu kerneEntry pointer via %rip, and we can remove the requirement of %gs, and remove the all the gr0 GS related code at all including the slowly swapgs. So there is possible that we won't have %GS pointing to any kernel object in future. It is better to make the check on golang code rather than asm code. vCPU.tid is other source of checking I think.

Reply:
Only a few limited sections of the kvm code can actually rely on being in guest mode (in the KernelException and KernelSyscall handlers, and between the entersyscall() and exitsyscall() bits in the SwitchToUser function). Typically you have a reference to the vCPU already, or one is passed from ring0.

Reply:
I don't get your request, should I add some check back?